core/src/cell/bag_of_cells.rs

```diff
@@ -67,18 +67,33 @@ impl BagOfCells {
 
     pub fn parse(serial: &[u8]) -> Result<BagOfCells, TonCellError> {
         let raw = RawBagOfCells::parse(serial)?;
-        let num_cells = raw.cells.len();
-        let mut cells: Vec<ArcCell> = Vec::with_capacity(num_cells);
+        let raw_cells_len = raw.cells.len();
+        let mut cells: Vec<ArcCell> = Vec::with_capacity(raw_cells_len);
 
         for (cell_index, raw_cell) in raw.cells.into_iter().enumerate().rev() {
             let mut references = Vec::with_capacity(raw_cell.references.len());
-            for ref_index in &raw_cell.references {
-                if *ref_index <= cell_index {
-                    return Err(TonCellError::boc_deserialization_error(
-                        "References to previous cells are not supported",
-                    ));
+            for ref_index in raw_cell.references {
```

Comment:
Some of these checks will be gone,when size in RawBagOfCells::parse gets a check that it is >0